### ISSUE TITLE
Enhancement to open multiple dialogs using $mdDialog

### ISSUE THREAD

**Issue by denver-HJS** [angular/material#8630]
Hello,

I am inquiring to find out whether or not there are plans to add functionality to the $mdDialog service for showing multiple dialogs. I am aware of [this thread](https://github.com/angular/material/issues/698) but I'd like to rekindle the discussion.

My specific need is that my application requires a virtual keyboard. I implemented this virtual keyboard using a wrapper service around $mdDialog (and a separate directive). This worked flawlessly...up until the point I tried to apply my virtual keyboard to an input that's inside of a dialog. Once the user clicks the input within the dialog it goes away and up comes the keyboard.

**1)** Is there any chance that having multiple dialogs will be implemented in the future?
**2)** Does anyone have suggestions for working around this limitation?

I found the options $mdDialog provides to be extremely useful, so it's disappointing to rearrange my architecture and try to recreate the niceties myself.

```
showKeyboard(keyboardType, inputElement, modelUpdateEvent, customValidation) {
      var s = this;
      let positionKeyboard = () => {
        $('#keyboard-dialog').position({
          of: $(inputElement),
          my: "right top",
          at: "right bottom",
          collision: "flipfit flipfit"
        });
        s._isDragging = false;
        let dragDelay = 1000;
        let dragTimeout = undefined;
        $('#keyboard-dialog').on("mousedown", () => {
          if(s._isDragging == false) {
            dragTimeout = this._$timeout(() => {
              s._isDragging = true;
              s._dragStartTime = moment();
              let draggingMask = $.parseHTML(`<div class="drag-mask"></div>`);
              $(draggingMask).css({
                position: 'absolute',
                width: '100%',
                height: '100%',
                top: '0px',
                left: '0px',
                zIndex: 100
              });
              $('#keyboard-dialog').append(draggingMask);
              $('#keyboard-dialog').css({
                border: '3px solid rgb(86, 180, 239)',
                'box-shadow': '0px 1px 3px rgba(0, 0, 0, 0.05) inset, 0px 0px 8px rgba(82, 168, 236, 0.6)'
              });
              $('#keyboard-dialog').draggable();
            }, dragDelay);
          }
        });

        $('#keyboard-dialog').on("mouseup", () => {
          if(angular.isDefined(dragTimeout)) {
            s._$timeout.cancel(dragTimeout);
            dragTimeout = undefined;
          }
          let now = moment();
          let timespan = moment.duration(now.diff(s._dragStartTime));
          if($(".drag-mask").length && s._isDragging && timespan.asMilliseconds() > 700) {
            s._isDragging = false;
            $(".drag-mask").remove();
            $('#keyboard-dialog').css({
              border: 'none',
              'box-shadow': 'none'
            });
            $('#keyboard-dialog').draggable("destroy");
          }
        });
      };

      this._$mdDialog.show({
        templateUrl: 'partials/dialogs/keyboard/keyboardDialog.html',
        controller: 'keyboardDialogController as vm',
        clickOutsideToClose: true,
        hasBackdrop: false,
        //openFrom: inputElement,
        closeTo: inputElement,
        onComplete: positionKeyboard,
        locals: {
          keyboardType: keyboardType,
          inputElement: inputElement,
          modelUpdateEvent: modelUpdateEvent,
          customValidation: customValidation
        }
      })
      .then(
        () => { $(inputElement).trigger("blur"); },
        () => { $(inputElement).trigger("blur"); }
      );
    }
```

**Comment by DevVersion** [comment]
We do not have plans to support it the new future. But we will create the base for multiple interim elements (like the dialog) in #8624. So once #8623 is ready, we'll probably revisit the multiple dialogs issue.

**Comment by denver-HJS** [comment]
Alright, understood. Thanks for the quick feedback =)

**Comment by jaymanned** [comment]
@DevVersion Will you let us know here if you finally manage to revisit the multiple dialogs issue?

**Comment by DevVersion** [comment]
@jaymanned Definitely yes! It's just a lower priority on my queue for now, but will definitely revisit that issue.

**Comment by olingern** [comment]
Copying and pasting my comment from #3072 for others who may stumble upon this or are unaware of existing functionality:
____
Unsure why this isn't documented by using `skipHide` on the child dialog will allow this. 

[Here's a plunker](http://plnkr.co/edit/Ga027OYU5nrkua3JxNRy?p=preview)

And [here's the line of code](https://github.com/angular/material/blob/e898d228dc6dd186b5b71cb1ea3d7cf2165528d5/src/core/services/interimElement/interimElement.js#L292) that "skips" hiding the parent dialog.

**Comment by DevVersion** [comment]
@olingern This is not documented, because it's currently only internal and designed for use with the `$mdMenu`.

In #9053, skipHide` has been redesigned to properly handle multiple interims at the same time.

**Comment by olingern** [comment]
@DevVersion I see... and awesome. Thanks for your work on this!

 So as of #9053 you migrate `skipHide` to `multiple` and keep a stack of shown interim elements?

**Comment by DevVersion** [comment]
@olingern - Exactly `skipHide` will be renamed to `multiple`.

- Once `multiple` is disabled (by default), the previous interim, when a new interim will be shown, will hide properly. (This was not the case before - overlapping toasts e.g)

- Keeping track of interims, which are currently `opening | open | closing`.
  This allows us to properly support cancellation / hiding of multiple interims.

**Comment by olingern** [comment]
Thanks for this. Super helpful for confirmations when inside a modal. I'm sure #698, #3072 will welcome the changes, too. Cheers.

**Comment by DevVersion** [comment]
Hello guys, just wanted to let you know, that multiple dialogs are right now supported, but not documented yet.

Using the `multiple: true` option does no longer close the previous showing dialog.

**Comment by jaymanned** [comment]
Great. Thanks. So it's now part of release 1.1.1?

**Comment by DevVersion** [comment]
It will be available in the next release. Right now its only in `master`.

**Comment by makkart** [comment]
Awesome, thank you!

**Comment by denver-HJS** [comment]
This is awesome news, thank you so much!

**Comment by warent** [comment]
Are we not able to use `.multiple(true)` on confirm dialogs?

**Comment by DevVersion** [comment]
@warent You should be able to do that.

**Comment by michaelcode** [comment]
Hello. Is this option part of release 1.1.1 yet?

Thx for the good work

**Comment by trinvh** [comment]
@DevVersion Not works with confirm().multiple(true) on 1.1.1

**Comment by DevVersion** [comment]
@trinvh It doesn't work because it hasn't been released as part of `v1.1.1`. The feature is included in `v.1.1.3`.

**Comment by trinvh** [comment]
Well done @DevVersion

**Comment by njcaruso** [comment]
This works for me:

var confirmDialog = $mdDialog.confirm({
  multiple: true
});

$mdDialog.show(confirmDialog)

**Comment by richardanmmol** [comment]
Great! It worked for me.

**Comment by jainpiyush111** [comment]
I am not able to use .multiple(true) on prompt dialogs?

**Comment by njcaruso** [comment]
@jainpiyush111  - I haven't tested on the prompt(), but on confirm() you would do it this way, which is probably very similar for prompt:

```
        var confirmDialog = $mdDialog.confirm()
            .title('Please confirm')
            .textContent('Are you sure you want to delete this message?')
            .ariaLabel('Delete confirmation dialog')
            .ok('Yes')
            .cancel('No')
            .multiple(true);

        $mdDialog.show(confirmDialog).then(function() {
```

**Comment by yabhis** [comment]
For confirm and alert dialog, it'll be better to use .parent with skipHide/multiple:
``` 
        var confirmDialog = $mdDialog.confirm({skipHide/multiple: true})
            .title('Please confirm')
            .textContent('Are you sure you want to delete this message?')
            .parent(angular.element(document.querySelector('#parentDialog')))
            .ok('Yes')
            .cancel('No')
            .multiple(true);

        $mdDialog.show(confirmDialog).then(function(){}
```

`skipHide` for older versions & `multiple` for newer...

**Comment by okaiyong** [comment]
I am using 1.1.4 and this is my codes:
$mdDialog.show(
  { multiple: true, ..... }
);

It didn't work, any reason?

**Comment by njcaruso** [comment]
Not sure about your version, but I'm on 1.1.7 and it works well.

**Comment by okaiyong** [comment]
Thanks. Solved.
